File: serdaSimus/assignments/02SlidingWindowPattern/lc209/lc209.cpp
using namespace std;

#include <iostream>
#include <limits>
#include <vector>

class MinSizeSubArraySum {
 public:
  static int findMinSubArray(int S, const vector<int>& arr) {
    // TODO: Write code here
  }
};

int main(int argc, char* argv[]) {
  int result = MinSizeSubArraySum::findMinSubArray(7, vector<int>{2, 1, 5, 2, 3, 2});
  cout << "Smallest subarray length: " << result << endl;
  result = MinSizeSubArraySum::findMinSubArray(7, vector<int>{2, 1, 5, 2, 8});
  cout << "Smallest subarray length: " << result << endl;
  result = MinSizeSubArraySum::findMinSubArray(8, vector<int>{3, 4, 1, 1, 6});
  cout << "Smallest subarray length: " << result << endl;
}







    // Solution
    // -----
    // int windowSum = 0, minLength = numeric_limits<int>::max();
    // int windowStart = 0;
    // for (int windowEnd = 0; windowEnd < arr.size(); windowEnd++) {
    //   windowSum += arr[windowEnd];  // add the next element
    //   // shrink the window as small as possible until the 'windowSum' is smaller than 'S'
    //   while (windowSum >= S) {
    //     minLength = min(minLength, windowEnd - windowStart + 1);
    //     windowSum -= arr[windowStart];  // subtract the element going out
    //     windowStart++;                  // slide the window ahead
    //   }
    // }

    // return minLength == numeric_limits<int>::max() ? 0 : minLength;

File: kelvinWu/slidingwindow/lc209/lc209.js
// Problem Statement #

// Given an array of positive numbers and a positive number ‘S,’ find the length of the smallest contiguous subarray whose sum is greater than or equal to ‘S’. Return 0 if no such subarray exists.

//keep track of two pointers, start and end
//move end pointer until you find valid window
//check if window size is smaller than the best, save it
//if you find a valid window, try to shrink by moving start until it becomes invalid
//repeat above steps

function smallest_subarray_with_given_sum(s, arr) {
  // TODO: Write code here
  let minLength = Number.MAX_SAFE_INTEGER;
  let start = 0,
    end = -1;
  let currentSum = 0;
  while (end < arr.length) {
    if (currentSum >= s) {
      currentSum -= arr[start];
      start++;
    } else {
      end++;
      if (end != arr.length) currentSum += arr[end];
    }
    // console.log(currentSum);
    if (currentSum >= s) minLength = Math.min(minLength, end - start + 1);
  }
  if (minLength == Number.MAX_SAFE_INTEGER) return 0;
  return minLength;
}

console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    7,
    [2, 1, 5, 2, 3, 2]
  )}`
);
console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    7,
    [2, 1, 5, 2, 8]
  )}`
);
console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    8,
    [3, 4, 1, 1, 6]
  )}`
);

// Solution
// -----
// let windowSum = 0,
// minLength = Infinity,
// windowStart = 0;

// for (windowEnd = 0; windowEnd < arr.length; windowEnd++) {
// windowSum += arr[windowEnd]; // add the next element
// // shrink the window as small as possible until the 'window_sum' is smaller than 's'
// while (windowSum >= s) {
//   minLength = Math.min(minLength, windowEnd - windowStart + 1);
//   windowSum -= arr[windowStart];
//   windowStart += 1;
// }
// }

// if (minLength === Infinity) {
// return 0;
// }
// return minLength;
